src/relations/hacluster.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""hacluster relation hooks and helpers."""
+
+import json
+import logging
+from hashlib import shake_128
+from ipaddress import IPv4Address, ip_address
+from typing import Optional, Union
+
+import ops
+
+HACLUSTER_RELATION_NAME = "ha"
+
+logger = logging.getLogger(__name__)
+
+
+class HACluster(ops.Object):
+    """Defines hacluster functunality."""
+
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, HACLUSTER_RELATION_NAME)
+
+        self.charm = charm
+
+        self.framework.observe(
+            charm.on[HACLUSTER_RELATION_NAME].relation_changed, self._on_changed
+        )
+        self.framework.observe(charm.on.config_changed, self._on_changed)
+
+    @property
+    def relation(self) -> ops.Relation:
+        """Returns the relations in this model, or None if hacluster is not initialised."""
+        return self.charm.model.get_relation(HACLUSTER_RELATION_NAME)
+
+    def _is_clustered(self) -> bool:
+        for key, value in self.relation.data.items():
+            if isinstance(key, ops.Unit) and key != self.charm.unit:
+                if value.get("clustered") in ("yes", "true"):
+                    return True
+                break
+        return False
+
+    def _on_changed(self, event: Union[ops.RelationChangedEvent, ops.RelationBrokenEvent]) -> None:
+        self.set_vip(self.charm.config.get("vip"))
+
+    def get_unit_juju_status(self) -> ops.StatusBase:
+        """Returns the status of the hacluster if relation exists."""
+        if self.relation and not self.charm.is_externally_accessible(event=None):
+            return ops.BlockedStatus("ha integration used without data-integrator")
+
+        vip = self.charm.config.get("vip")
+        if self.relation and not vip:
+            return ops.BlockedStatus("ha integration used without vip configuration")
+
+        if vip and not self.charm.is_externally_accessible(event=None):
+            return ops.BlockedStatus("vip configuration without data-integrator")
+
+        if self.charm.is_workload_authenticated and self.charm.unit.is_leader() and vip:
+            return ops.ActiveStatus(f"VIP: {vip}")
+
+    def set_vip(self, vip: Optional[str]) -> None:
+        """Adds the requested virtual IP to the integration."""
+        if not self.relation:
+            return
+
+        if not self._is_clustered():
+            logger.debug("early exit set_vip: ha relation not yet clustered")
+            return
+
+        if vip:
+            # TODO Add nic support
+            ipaddr = ip_address(vip)
```

Comment:
the validation story is even more complex - currently we require that provided IP address is in the same subnet as the machines of the model (otherwise pacemaker is unable to create/manage the VIP and it borks the integration with hacluster)

i understand that the concern is security related - the library that converts the VIP into an ip address already has basic validation:
```
>>> from ipaddress import ip_address
>>> ip_address('a.b.c.d')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/ipaddress.py", line 54, in ip_address
    raise ValueError(f'{address!r} does not appear to be an IPv4 or IPv6 address')
ValueError: 'a.b.c.d' does not appear to be an IPv4 or IPv6 address
```